simplex/engine.go

```diff
@@ -0,0 +1,172 @@
+	"fmt"
+	"time"
+
+	"github.com/ava-labs/simplex"
+	"github.com/ava-labs/simplex/wal"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/proto/pb/p2p"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/logging"
+)
+
+var _ common.Engine = (*Engine)(nil)
+
+var (
+	maxProposalWaitTime = time.Second * 2
+	maxRebroadcastWait  = time.Second * 2
+
+	errUnknownMessageType = errors.New("unknown message type")
+	errConvertingMessage  = errors.New("error converting message to simplex message")
+)
+
+type Engine struct {
+	common.Handler
+
+	epoch              *simplex.Epoch
+	blockDeserializer  *blockDeserializer
+	quorumDeserializer *QCDeserializer
+	logger             logging.Logger
+}
+
+// THe VM must be initialized before creating the engine
+func NewEngine(ctx context.Context, config *Config) (*Engine, error) {
+	signer, verifier := NewBLSAuth(config)
+	qcDeserializer := &QCDeserializer{
+		verifier: &verifier,
+	}
+	blockDeserializer := &blockDeserializer{
+		parser: config.VM,
+	}
+	signatureAggregator := &SignatureAggregator{
+		verifier: &verifier,
+	}
+	comm, err := NewComm(config)
+	if err != nil {
+		return nil, err
+	}
+
+	storage, err := newStorage(ctx, config, qcDeserializer, &blockTracker{})
+	if err != nil {
+		return nil, err
+	}
+	wal, err := wal.New(config.WalLocation)
+	if err != nil {
+		return nil, err
+	}
+
+	lastBlock, _, err := storage.Retrieve(storage.NumBlocks() - 1)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't find last block at height %d: %w", storage.NumBlocks()-1, err)
+	}
+
+	blockTracker := newBlockTracker(lastBlock.(*Block))
+
+	blockBuilder := &BlockBuilder{
+		vm:           config.VM,
+		blockTracker: blockTracker,
+		log:          config.Log,
+	}
+
+	epochConfig := simplex.EpochConfig{
+		MaxProposalWait:     maxProposalWaitTime,
+		MaxRebroadcastWait:  maxRebroadcastWait,
+		QCDeserializer:      qcDeserializer,
+		Logger:              config.Log,
+		ID:                  config.Ctx.NodeID[:],
+		Signer:              &signer,
+		Verifier:            &verifier,
+		BlockDeserializer:   blockDeserializer,
+		SignatureAggregator: signatureAggregator,
+		Comm:                comm,
+		Storage:             storage,
+		WAL:                 wal,
+		BlockBuilder:        blockBuilder,
+		Epoch:               0, // 0 for now, but we would get the epoch from the metadata associated with the latest block
+		StartTime:           time.Now(),
+		ReplicationEnabled:  true,
+	}
+
+	epoch, err := simplex.NewEpoch(epochConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Engine{
+		epoch:              epoch,
+		blockDeserializer:  blockDeserializer,
+		quorumDeserializer: qcDeserializer,
+		logger:             config.Log,
+	}, nil
+}
+
+func (e *Engine) Start(_ context.Context, _ uint32) error {
+	e.logger.Info("Starting simplex engine")
+	return e.epoch.Start()
+}
+
+func (e *Engine) SimplexMessage(nodeID ids.NodeID, msg *p2p.Simplex) error {
```

Comment:
This is the wrong method signature isn't it? 

I think it should be `ctx context.Context, nodeID ids.NodeID, msg *p2ppb.Simple`.

This also makes me think that ideally this PR should be accompanied by an e2e integration test that can spin up Simplex processes. Otherwise we never know if all the glue code actually works properly.